File: src/com/anathema_roguelike/actors/Actor.java
package com.anathema_roguelike.actors;

import com.anathema_roguelike.entities.characters.actions.costs.EnergyCost;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public interface Actor {
	
	Duration getDuration();
	Energy getEnergy();
	Optional<Action<?>> getNextAction();

	default Action<Actor> getDefaultAction() {
		return new Action<>(this, EnergyCost.STANDARD(this)) {
			@Override
			protected void onTake() {

			}
		};
	}
	
	default void energize() {
		getEnergy().energize();
	}
	
	default double getEnergyLevel() {
		return getEnergy().getEnergyLevel();
	}
	
	default void act() {
		getNextAction().ifPresentOrElse(a -> a.take(), () -> getDefaultAction().take());
	}
}

File: src/com/anathema_roguelike/actors/Actor.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package actors

import com.anathema_roguelike.entities.characters.actions.costs.EnergyCost
import com.anathema_roguelike.main.Game

trait Actor {

  private val energy = new Energy

  Game.getInstance.getState.registerActor(this)

  def getDuration: Duration

  def getNextAction: Option[Action[_]]

  def getDefaultAction: Action[Actor] = new Action[Actor](this, EnergyCost.STANDARD(this)) {
    override protected def onTake(): Unit = {

    }
  }

  def getEnergy: Energy = energy

  def energize(): Unit = {
    getEnergy.energize()
  }

  def getEnergyLevel: Double = getEnergy.getEnergyLevel

  def act(): Unit = {
    getNextAction.getOrElse(getDefaultAction).take()
  }
}
